src/plugins/data/server/query/route_handler_context.ts

```diff
@@ -72,26 +73,62 @@ function extractReferences({
   return { attributes, references };
 }
 
+function createBadRequestError(message: string): SavedQueryRestResponse {
+  return {
+    status: 400,
+    body: {
+      message,
+    },
+  };
+}
+
 function verifySavedQuery({ title, query, filters = [] }: SavedQueryAttributes) {
   if (!isQuery(query)) {
-    throw new Error(`Invalid query: ${query}`);
+    return createBadRequestError(`Invalid query: ${JSON.stringify(query, null, 2)}`);
   }
 
   if (!isFilters(filters)) {
-    throw new Error(`Invalid filters: ${filters}`);
+    return createBadRequestError(`Invalid filters: ${JSON.stringify(filters, null, 2)}`);
   }
 
   if (!title.trim().length) {
-    throw new Error('Cannot create saved query without a title');
+    return createBadRequestError('Cannot create query without a title');
   }
 }
 
 export async function registerSavedQueryRouteHandlerContext(context: RequestHandlerContext) {
   const soClient = (await context.core).savedObjects.client;
 
-  const createSavedQuery = async (attrs: SavedQueryAttributes) => {
-    verifySavedQuery(attrs);
+  const validateSavedQueryTitle = async (attributes: InternalSavedQueryAttributes, id?: string) => {
+    const { savedQueries } = await findSavedQueries({
+      page: 1,
+      perPage: 1,
+      search: `"${escapeKuery(attributes.title.trim())}"`,
+    });
+    const existingQuery = savedQueries[0];
```

Comment:
Since this check isn't using the `titleKeyword` field, shouldn't we be checking all the values inside `savedQueries` and not just the first?

Reply:
Thinking more about this, we might want to actually provide another utility that actually checks the `titleKeyword` since we're wanting an exact match against the term, and you could theoretically get an infinite number of matches against the text version (because the text version is analyzed and special characters are discarded).

To illustrate, create a saved query with title `foo bar`, then create a second one with `foo* bar`. Then, try to create another one with `foo* bar` (it will let you).

Happy to discuss this over zoom if this isn't very clear.

Reply:
These issues should also be taken care of now with the new `_is_duplicate_title` endpoints.